### PR TITLE
Connect: Make `tsh ssh` in command bar not depend on resource cache

### PR DESCRIPTION
**Pull request by ravicious**
Part of #21964.

Executing `tsh ssh` from the command bar depends on the resource cache as it needs to translate the resource hostname from the command (e.g. `tsh ssh root@ubuntu`) into a server UUID that's necessary to open a new terminal tab.

Since we want to remove the cache, there's no way of obtaining the UUID from the hostname other than hitting the proxy. Which is what this PR accomplishes.

It does so in a rather naive way. The next PR is going to move the whole logic of resolving the hostname from the command launcher to `useDocumentTerminal`. This way the `tsh ssh` command in the command bar itself will stay synchronous and all the waiting will be done in `DocumentTerminal`.

If you want to check how the final "product" feels like, check out the branch of #22182.

### CODE REVIEW

**Comment by ravicious** [comment]
@gzdunek @avatus I wanted to split the four PRs between you two, but so far I only published two because I need your input on this.

Initially I did what's in this PR: `tsh ssh` executed from the command bar will now reach out to the proxy to translate server hostname into a UUID. This however means that after you type `tsh ssh root@ubuntu` and press enter, the command will not execute immediately because we need to wait for the response from the proxy. So I made another PR (#21977) on top of the changes from this one which disables the input and shows a spinner until we get the UUID.

While publishing this PR, I realized that we could move this logic to `useDocumentTerminal` instead and this would keep command bar commands sync again, see the last section of the description.

What do you say about that? I feel like it's worth it to reexamine the approach and move this logic to `useDocumentTerminal`. Overall this will result in a better UX and all we have to do is to move stuff to `useDocumentTerminal` and add a new field on `DocumentTshNode`.

**Comment by avatus** [comment]
I'll take a look at this when I return to the office on tuesday

**Comment by ravicious** [comment]
Ah, I forgot about the bank holiday! By this time though I'll most likely have another PR up which demonstrates how this will work after moving it to `useDocumentTerminal`. I strongly believe this is the right thing to do, as adding a loading state to the command bar right after the user tries to execute the command makes less sense than doing it within `useDocumentTerminal`.

**Comment by gzdunek** [comment]
@ravicious I just tried this and having to wait for the tab to open feels a bit unresponsive.
Moving the loading state to `DocumentTerminal` seems like a much better idea to me :)

**Comment by ravicious** [comment]
Thanks for the feedback, I addressed this in the other PRs (#22181 and #22182).

This PR will still be merged in even though #22181 will move resolving the hostname from the command launcher to `useDocumentTerminal`. This way we can review the changes in the logic separately from the action of moving it from one place to another.

File: web/packages/teleterm/src/ui/commandLauncher.ts [hunk 1]
```diff
@@ -55,23 +57,30 @@ const commands = {
...
+        if (error instanceof AmbiguousHostnameError) {
+          // Log the ambiguity of the hostname but continue anyway. This will pass the ambiguous
+          // hostname to tsh ssh and show an appropriate error in the new tab.
+          console.error(error.message);
+        } else {
+          ctx.notificationsService.notifyError({
+            title: `Could not establish a connection to ${host}`,
+            description: error['message'],
+          });
+          return;
```
Comment (avatus) [inline web/packages/teleterm/src/ui/commandLauncher.ts:80]: Do I understand this correctly? If its an ambiguoushostnameerror we want to let the code continue firing after the console log, but other errors stop the rest of the execution? What is the reasoning behind "allowing" this type of error to be shown in a new tab and not the rest?

File: web/packages/teleterm/src/ui/uri.ts [hunk 2]
```diff
@@ -23,6 +23,8 @@ export const paths = {
   leafCluster: '/clusters/:rootClusterId/leaves/:leafClusterId',
   server:
     '/clusters/:rootClusterId/(leaves)?/:leafClusterId?/servers/:serverId',
+  serverLeaf:
```
Comment (avatus) [inline web/packages/teleterm/src/ui/uri.ts:26]: should we keep the name consistent with leaf first? `leafServer`? I see instances of `leafServer` in the codebase but no `serverLeaf`

**Review by avatus — APPROVED** [review]
Approved with small questions!

File: web/packages/teleterm/src/ui/commandLauncher.ts (hunk 1 above)
Comment (ravicious) [reply web/packages/teleterm/src/ui/commandLauncher.ts:80]: The original idea when I first implemented that was that we allow `AmbiguousHostnameError` to pass through because we're essentially shifting the responsibility from Connect to tsh. We know that if we attempt to execute `tsh ssh` with the given login and hostname, tsh itself will display an appropriate error so we let it do that as at the time we didn't have a good mode of showing this error in any other way in Connect – our only other option was showing an error notification which we deemed subpar to a failed `tsh ssh` execution in a PTY.

So in the end, we do that because we didn't (have time to) think of a better way to approach this particular failure.

Other errors are not passed through because `resourcesService.getServerByHostname` makes a network call so there's a myriad of things that can go wrong and we cannot continue if we get an unexpected error here.

In other words, `AmbiguousHostnameError` being thrown by `ResourcesService` directly maps to a certain error handling scenario of `tsh ssh`. We cannot say that of other errors that might be thrown by `ResourcesService`.

File: web/packages/teleterm/src/ui/uri.ts (hunk 2 above)
Comment (ravicious) [reply web/packages/teleterm/src/ui/uri.ts:26]: I think I just wanted the key to start with the most important descriptor in this set of keys (what kind of resource this path is for) and only then followed by a minor descriptor of whether it's for a leaf resource or not. So it's referring less to the concept of a leaf server and more to the concept of a path for servers that are leaf servers.

Alas, we already have `rootCluster` and `leafCluster` within the same object, I guess I could rename then to `clusterRoot` and `clusterLeaf` if I wanted to be consistent. 🤔

Comment (avatus) [reply web/packages/teleterm/src/ui/uri.ts:26]: I see. I really like that idea of not describing the "item" but describing the path and it makes _much_ more sense as a path descriptor to follow your `serverLeaf` convention. feels very REST-y. I'm ok with keeping it then and, if you feel compelled to change the others thats fine too but, either way, I see the light now. Thanks for the insight

**Comment by ravicious** [comment]
Merged into #21964 in 7eaa33f.
